onnx/defs/traditionalml/defs.cc

```diff
@@ -314,103 +314,147 @@ static const char* LabelEncoder_ver2_doc = R"DOC(
+            AttributeProto::TENSOR,
+            OPTIONAL_VALUE)
         .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
-          // Label encoder is one-to-one mapping.
-          if (ctx.getNumInputs() != 1) {
-            fail_shape_inference("Label encoder has only one input.");
+          // Ensure only one keys_* attribute is set, and get the corresponding attribute type
+          int32_t key_type = TensorProto_DataType_UNDEFINED;
+          const AttributeProto* keys_as_tensor = ctx.getAttribute("keys_as_tensor");
+          if (keys_as_tensor != nullptr) {
+            // Ensure keys_as_tensor is 1D
+            if (keys_as_tensor->t().dims_size() != 1) {
+              fail_shape_inference("keys_as_tensor must be 1D.");
+            }
+            key_type = keys_as_tensor->t().data_type();
           }
-          if (ctx.getNumOutputs() != 1) {
-            fail_shape_inference("Label encoder has only one output.");
+          const AttributeProto* keys_strings = ctx.getAttribute("keys_strings");
+          if (keys_strings != nullptr) {
+            key_type = TensorProto_DataType_STRING;
           }
+          const AttributeProto* keys_int64s = ctx.getAttribute("keys_int64s");
+          if (keys_int64s != nullptr) {
+            key_type = TensorProto_DataType_INT64;
+          }
+          const AttributeProto* keys_floats = ctx.getAttribute("keys_floats");
+          if (keys_floats != nullptr) {
+            key_type = TensorProto_DataType_FLOAT;
+          }
+          if (static_cast<int>(keys_as_tensor != nullptr) + static_cast<int>(keys_strings != nullptr) +
+                  static_cast<int>(keys_int64s != nullptr) + static_cast<int>(keys_floats != nullptr) !=
+              1) {
+            fail_shape_inference(
+                "One and only one of keys_as_tensor, keys_strings, keys_int64s, keys_floats must be set.");
+          }
+          if (key_type != ctx.getInputType(0)->tensor_type().elem_type()) {
+            fail_shape_inference(
+                "The input the key types must be identical but were ",
+                ctx.getInputType(0)->tensor_type().elem_type(),
+                " and ",
+                key_type,
+                " respectively.");
+          }
+
+          // Ensure only one values_* attribute is set, and get the corresponding attribute type
+          int32_t value_type = TensorProto_DataType_UNDEFINED;
+          const AttributeProto* values_as_tensor = ctx.getAttribute("values_as_tensor");
+          if (values_as_tensor != nullptr) {
+            // Ensure values_as_tensor is 1D
+            if (values_as_tensor->t().dims_size() != 1) {
+              fail_shape_inference("values_as_tensor must be 1D.");
+            }
+            if (keys_as_tensor == nullptr) {
+              fail_shape_inference("keys_as_tensor must be set if values_as_tensor is set.");
+            }
+            if (keys_as_tensor->t().dims_size() != values_as_tensor->t().dims_size()) {
```

Comment:
I guess you mean `dims(0)` instead of `dims_size()`, since the rank has already been verified to be 1.